### ISSUE TITLE
Migrate away from using fxa-auth-db's database & schema

### ISSUE THREAD

**Issue by jbuck** [mozilla/fxa-email-service#166]
There's two parts to this:

1) From an operational POV, once we're only using fxa-email-service in production we can spin up a new database rather than reusing the fxa-auth-db's database.
2) It'd be good to add database migrations to this repo at that point in time

**Comment by philbooth** [comment]
Moved this into the milestone I'm using for Q4 work.

**Comment by jbuck** [comment]
Redis, MySQL, and PostgreSQL are all supported on GCP. I'd recommend simplicity and choose one data store for all types of data. Redis feels like the right choice to me given the requirements, though it does mean we'd need to migrate data from MySQL to Redis. Not the end of the world, but worth noting.

**Comment by philbooth** [comment]
I'm definitely happy with Redis, although I've noticed that when I suggest it for FxA stuff people often pull a funny face at me and say something like "we need a proper database for this" <sup>[citation needed]</sup>.

So, just checking (and I don't really know what I'm talking about), are we happy with the consistency / persistence guarantees with Redis? I guess bounce records and metrics data are both ephemeral to some extent anyway so it's not the end of the world if we were to lose some, but the configuration data not so much.

Also, just linking to earlier discussion about this, in #4 @rfk cautioned that some of the data will be sensitive and that we should keep that in mind when picking a data store.

**Comment by jbuck** [comment]
If the auth server needs to store relational data than MySQL or Postgres totally make sense, but from your description of the data being stored, it sounds like Redis is a good fit. Maybe post your ideal structs for each of the data sets (bounce records, config, metrics) in this bug and we can decide for sure?

I'm happy with the consistency guarantees of Redis, as it's a single-threaded service so everything is atomic by default. The persistence guarantees I'm happy enough so long as we use a highly-available setup in prod, which we do for all of our Redis instances so far.

In terms of sensitive data we follow the principle of least-access so only the service, @jrgm, and I should have access

**Comment by jbuck** [comment]
If the auth server needs to store relational data than MySQL or Postgres totally make sense, but from your description of the data being stored, it sounds like Redis is a good fit. Maybe post your ideal structs for each of the data sets (bounce records, config, metrics) in this bug and we can decide for sure?

I'm happy with the consistency guarantees of Redis, as it's a single-threaded service so everything is atomic by default. The persistence guarantees I'm happy enough so long as we use a highly-available setup in prod, which we do for all of our Redis instances so far.

In terms of sensitive data we follow the principle of least-access so only the service, @jrgm, and I should have access

**Comment by philbooth** [comment]
> If the auth server needs to store relational data than MySQL or Postgres totally make sense, but from your description of the data being stored, it sounds like Redis is a good fit. Maybe post your ideal structs for each of the data sets (bounce records, config, metrics) in this bug and we can decide for sure?

I am more than happy with Redis fwiw, I just wanted to make sure this issue captured all of the discussion that had taken place elsewhere (and give anyone else the chance to comment).

There is one relational aspect to the data as I see it right now, but it's only a single property so we can just duplicate it in both data types when writing and then there's no need to look up the related type.

These are the data structures that I expect us to store:

* Bounce record (no relations, indexed by email address):

  ```json
  [
    {
      "bounce_type": 1,
      "bounce_subtype": 1,
      "created_at": 1538040018453
    }
  ]
  ```

* Config (no relations, indexed by client id, `providers` part is still quite hand-wavy):

  ```json
  {
    "notification_queue": "https://sqs.us-east-1.amazonaws.com/927034868273/fxa-email-notification-dev",
    "providers": {
      "ses": {
        "priority": 1
      },
      "sendgrid": {
        "priority": 2,
        "try_bounces": true,
        "regex": ".+@hotmail\\.fr$"
      }
    }
  }
  ```

* Metrics (relation with `config.notification_queue`, indexed by message id):

  ```json
  {
    "flow_id": "c75b9b15b81aff8a2081dce38871416fa7c36deeaa482ba53dafe4f05e2b0798",
    "flow_begin_time": 1538041898168,
    "browser": "Firefox",
    "browser_version": "64.0a1",
    "os": "MacOS",
    "os_version": "10.13.6",
    "device_type": "desktop",
    "entrypoint": "menupanel",
    "service": "sync",
    "utm_source": "foo",
    "notification_queue": "https://sqs.us-east-1.amazonaws.com/927034868273/fxa-email-notification-dev"
  }
  ```

And just to reiterate, I'm totes happy to denormalise `notification_queue` when writing the metrics data.

**Comment by philbooth** [comment]
I think there's three parts to this:

1. Patch the email service to use it's own data store. (easy)

2. Come up with a migration plan for copying existing bounce records to the new store and keeping the two stores in sync for the duration of the switchover in prod. (really, really hard)

3. Patch the email service to delete old bounce records from its store once we're happy everything works like clockwork. (easy)

**Comment by philbooth** [comment]
> 2. Come up with a migration plan for copying existing bounce records to the new store and keeping the two stores in sync for the duration of the switchover in prod. (really, really hard)

Maybe the best way to approach this is to make the first cut of the email service write to both data stores and read just from it's own. Then when we're confident about not wanting to go back to the auth server, we can stop it writing there. Does that make sense?

**Comment by philbooth** [comment]
(or even make the first cut write-to-both-read-from-authdb, the second cut write-to-both-read-from-own-db and then the third cut write-to-own-db-read-from-own-db)

**Comment by philbooth** [comment]
I've been implementing this today and a question has occurred to me. Do we consider email addresses sensitive enough to hash them before using as a key for bounce records?

I know we don't hash them in the auth db, but as we're writing this new service I thought it might be worth revisiting. Obviously the advantage to not hashing them is that it makes it easier to look them up when users complain about not receiving emails. But we could also write some tooling around that if it's the only reason.

Anyone have an opinion?

**Comment by philbooth** [comment]
> I've been implementing this today...

For the record, I've started off down the write-to-both-read-from-authdb for an initial cut.

**Comment by jbuck** [comment]
I agree that doing a dual write, single read is the best migration path forwards

**Comment by philbooth** [comment]
Although I've been working on this and have a patch that's nearly ready for review, there's some ops-related stuff that needs to happen before things can take effect in prod:

* Set up notification queues in dev, stage and prod for communication between the email service and the auth server. This is so that, when the email service starts processing bounce, complaint and delivery events, it can tell the auth server to log metrics (which is FxA-specific logic).

* Tweak the auth server so that it listens to the new notification queue and doesn't do any of the db-related activity associated with handling bounce, complaint and delivery events. Instead it will just log metrics data.

* Deploy the queues binary from this repo and configure it to point at the notification queue.

Some of these things need to happen simultaneously and some are pre-requisites for others, so I'll need to co-ordinate with @jrgm when I get back from PTO about when the convenient time to do it all is, what order to do it in and how we can test whether it's working when it's done.

Just wanted to get it written down somewhere now as I realised it's only in my head at that moment.

**Comment by vladikoff** [comment]
from mtg: @philbooth still 126?

**Comment by philbooth** [comment]
Bumping to train-127 because of the dependency on #186.
